packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/editor/EditorToolbarDecorator.kt

```diff
@@ -44,71 +39,38 @@ class EditorToolbarDecorator(
-                        .build(project, dataSource)
-                        .setRequestor(ConnectionRequestor.Anonymous())
-                        .setAskPassword(true)
-                        .setRunConfiguration(
-                            ConsoleRunConfiguration.newConfiguration(project).apply {
-                                setOptionsFromDataSource(dataSource)
-                            },
-                        )
-
-                val connectionJob = runCatching { connectionHandler.create()?.get() }
-                connectionJob.onFailure {
-                    toolbar.connecting = false
-                    toolbar.failedConnection = dataSource
-                    return@launchChildBackground
-                }
-
-                val connection = connectionJob.getOrNull()
-                toolbar.connecting = false
-
-                // could not connect, do nothing
-                if (connection == null || !dataSource.isConnected()) {
-                    toolbar.selectedDataSource = null // remove data source because we didn't connect
-                    return@launchChildBackground
-                }
-
-                editor.virtualFile?.putUserData(Keys.attachedDataSource, dataSource)
-                ApplicationManager.getApplication().invokeLater {
-                    val session = DatabaseSessionManager.openSession(editor.project!!, dataSource, null)
-                    val probe = NewConnectionActivatedProbe()
-                    probe.connected(session)
-
-                    val psiFile =
-                        PsiManager.getInstance(project).findFile(editor.virtualFile)
-                            ?: return@invokeLater
-                    DaemonCodeAnalyzer.getInstance(project).restart(psiFile)
-                }
-            }
-        } else {
-            editor.virtualFile?.putUserData(Keys.attachedDataSource, dataSource)
-            val psiFile = PsiManager.getInstance(project).findFile(editor.virtualFile) ?: return
-            DaemonCodeAnalyzer.getInstance(project).restart(psiFile)
+    fun onConnected(dataSource: LocalDataSource) {
+        editor.virtualFile?.putUserData(Keys.attachedDataSource, dataSource)
+        editor.virtualFile?.removeUserData(Keys.attachedDatabase)
+
+        ApplicationManager.getApplication().invokeLater {
+            val project = editor.project!!
+            val session = DatabaseSessionManager.openSession(project, dataSource, null)
+            val probe = NewConnectionActivatedProbe()
+            probe.connected(session)
+
+            analyzeFileFromScratch()
         }
     }
 
-    fun onDataSourceUnselected() {
+    fun onDisconnected() {
         editor.virtualFile?.removeUserData(Keys.attachedDataSource) ?: return
-        val psiFile = PsiManager.getInstance(editor.project!!).findFile(editor.virtualFile) ?: return
-        DaemonCodeAnalyzer.getInstance(editor.project).restart(psiFile)
+        editor.virtualFile?.removeUserData(Keys.attachedDatabase)
```

Comment:
We should probably remove both the keys and then return if applicable?

Reply:
Ah true, fixing it!